File: examples/ospf-kotlin-example/src/main/fuookami/ospf/kotlin/example/framework_demo/demo1/bandwidth_context/Aggregation.kt
package fuookami.ospf.kotlin.example.framework_demo.demo1.bandwidth_context

import fuookami.ospf.kotlin.utils.functional.*
import fuookami.ospf.kotlin.core.frontend.model.mechanism.*
import fuookami.ospf.kotlin.example.framework_demo.demo1.bandwidth_context.model.EdgeBandwidth
import fuookami.ospf.kotlin.example.framework_demo.demo1.bandwidth_context.model.NodeBandwidth
import fuookami.ospf.kotlin.example.framework_demo.demo1.bandwidth_context.model.ServiceBandwidth

class Aggregation(
    val edgeBandwidth: EdgeBandwidth,
    val serviceBandwidth: ServiceBandwidth,
    val nodeBandwidth: NodeBandwidth
) {
    fun register(model: LinearMetaModel): Try {
        val subprocesses = arrayListOf(
            { return@arrayListOf edgeBandwidth.register(model) },
            { return@arrayListOf serviceBandwidth.register(model) },
            { return@arrayListOf nodeBandwidth.register(model) }
        )

        for (subprocess in subprocesses) {
            when (val result = subprocess()) {
                is Failed -> {
                    return Failed(result.error)
                }

                is Ok -> {}
            }
        }
        return ok
    }
}

File: ospf-rust-example/src/framework/demo1/bandwidth_context/aggregation.rs
use super::model::{EdgeBandwidth, NodeBandwidth, ServiceBandwidth};

pub struct Aggregation {
    pub edge_bandwidth: EdgeBandwidth,
    pub service_bandwidth: ServiceBandwidth,
    pub node_bandwidth: NodeBandwidth,
}

impl Aggregation {
    pub fn new(
        edge_bandwidth: EdgeBandwidth,
        service_bandwidth: ServiceBandwidth,
        node_bandwidth: NodeBandwidth,
    ) -> Self {
        Self {
            edge_bandwidth,
            service_bandwidth,
            node_bandwidth,
        }
    }
}
